Mage.Sets/src/mage/cards/t/TheMillenniumCalendar.java

```diff
@@ -0,0 +1,150 @@
+        this.addAbility(ability);
+
+        // When there are 1,000 or more time counters on The Millennium Calendar, sacrifice it and each opponent loses 1,000 life.
+        this.addAbility(new TheMillenniumCalendarStateTriggeredAbility());
+    }
+
+    private TheMillenniumCalendar(final TheMillenniumCalendar card) {
+        super(card);
+    }
+
+    @Override
+    public TheMillenniumCalendar copy() {
+        return new TheMillenniumCalendar(this);
+    }
+}
+
+class TheMilleniumCalendarTriggeredAbility extends TriggeredAbilityImpl {
+
+    public TheMilleniumCalendarTriggeredAbility() {
+        super(Zone.BATTLEFIELD, null, false);
+    }
+
+    protected TheMilleniumCalendarTriggeredAbility(final TheMilleniumCalendarTriggeredAbility ability) {
+        super(ability);
+    }
+
+    @Override
+    public TheMilleniumCalendarTriggeredAbility copy() {
+        return new TheMilleniumCalendarTriggeredAbility(this);
+    }
+
+    @Override
+    public boolean checkEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.UNTAPPED_BATCH;
+    }
+
+    @Override
+    public boolean checkTrigger(GameEvent event, Game game) {
+        if (!game.isActivePlayer(getControllerId())) {
+            return false;
+        }
+        UntappedBatchEvent batchEvent = (UntappedBatchEvent) event;
+        int count = batchEvent
+                .getEvents()
+                .stream()
+                .filter(UntappedEvent::isAnUntapStepEvent)
+                .map(UntappedEvent::getTargetId)
+                .map(game::getPermanent)
+                .filter(Objects::nonNull)
+                .filter(p -> p.getControllerId().equals(getControllerId()))
+                .mapToInt(p -> 1)
+                .sum();
+
+        if (count <= 0) {
+            return false;
+        }
+
+        this.getEffects().clear();
+        this.addEffect(new AddCountersSourceEffect(CounterType.TIME.createInstance(count)));
+        this.getEffects().setValue(infoKey, count); // for the trigger on the stack to display the info.
+        return true;
+    }
+
+    private static final String infoKey = "number_untapped";
+
+    @Override
+    public String getRule() {
+        // this trigger might want to expose extra info on the stack
+        String triggeredInfo = "";
+        if (!this.getEffects().isEmpty()) {
+            triggeredInfo += "<br><br><i>Number of untapped permanents: " + this.getEffects().get(0).getValue(infoKey) + "</i>";
+        }
+        return "Whenever you untap one or more permanents during your untap step, "
+                + "put that many time counters on {this}."
+                + triggeredInfo;
+    }
+}
+
+/**
+ * Inspired by {@link mage.cards.n.NineLives}
+ */
+class TheMillenniumCalendarStateTriggeredAbility extends StateTriggeredAbility {
+
+    public TheMillenniumCalendarStateTriggeredAbility() {
+        super(Zone.BATTLEFIELD, new SacrificeSourceEffect().setText("sacrifice it"));
```

Comment:
another (new-ish) way to do this, rather than setting text on the effect, is to include within the trigger constructor:
``` this.replaceRuleText = true;```

Reply:
Oh that's the `{this} -> it` replacement in triggered effects' rules text I was missing out. Thanks for pointing that out.